lib/IRGen/GenProto.cpp

```diff
@@ -1022,25 +1022,91 @@ class irgen::ConformanceInfo {
   getConditionalConformances() const = 0;
 };
 
-static llvm::Value *
-emitWitnessTableAccessorCall(IRGenFunction &IGF,
-                             const NormalProtocolConformance *conformance,
-                             CanType conformingType,
-                             llvm::Value **srcMetadataCache) {
-  auto accessor =
-    IGF.IGM.getAddrOfWitnessTableAccessFunction(conformance, NotForDefinition);
+static std::pair<llvm::Value *, llvm::Value *>
+emitConditionalConformancesBuffer(IRGenFunction &IGF,
+                                  const ProtocolConformance *conformance) {
+  const SpecializedProtocolConformance *specialized = nullptr;
+
+  // Find the closest containing SpecializedProtocolConformance, since we need
+  // the substitution map to be able to find the ProtocolConformances for each
+  // conditional conformance. Conformances that aren't specialized (i.e. fully
+  // concrete ones) shouldn't be generic and so shouldn't have any conditional
+  // requirements.
+  while (true) {
+    if (isa<NormalProtocolConformance>(conformance)) {
+      break;
+    } else if (auto s = dyn_cast<SpecializedProtocolConformance>(conformance)) {
+      specialized = s;
+      break;
+    } else if (auto i = dyn_cast<InheritedProtocolConformance>(conformance)) {
+      conformance = i->getInheritedConformance();
+    }
+  }
+
+  // Pointers to the witness tables, in the right order, which will be included
+  // in the buffer that gets passed to the witness table accessor.
+  llvm::SmallVector<llvm::Value *, 4> tables;
+
+  if (specialized) {
+    auto subMap = specialized->getSubstitutionMap();
+
+    for (auto req : conformance->getConditionalRequirements()) {
+      if (req.getKind() != RequirementKind::Conformance)
+        continue;
+
+      auto type = req.getFirstType()->getCanonicalType();
+      auto proto = req.getSecondType()->castTo<ProtocolType>()->getDecl();
+
+      auto reqConformance = subMap.lookupConformance(type, proto);
```

Comment:
`type` here could be concrete, which is why I've hacked SubstitutionMap/GenericSignature to support pulling those conformances from the `SubstitutionList` into the generic signature too.

One alternative I just thought of is to walk through the conditional requirements on the normal conformance, so that the requirements themselves phrased in terms of generic parameters, and then use the substitution map to convert these into concrete types and also this call will work fine without needing hacks elsewhere.